ipn/ipnlocal: determine client-side-reachability by pinging

sfllaw opened the pull request:
PR #17367 introduced the `client-side-reachability` node attribute that switches the node to completely ignore the online signal from control. This was to work around the times when the control plane determines that a node is not online, while the node is still able to connect to its peers. However, this breaks high-availability failover, since nodes that are actually offline are never treated that way.

This PR makes nodeBackend.PeerIsReachable actually test reachability by disco-pinging the peer node.

In the future, the client itself should collect reachability data from active Wireguard flows.

Updates #17366
Updates tailscale/corp#30379
Updates tailscale/corp#32686
Updates tailscale/corp#33033

sfllaw committed:
ipn/ipnlocal: determine client-side-reachability by pinging

PR #17367 introduced the `client-side-reachability` node attribute
that switches the node to completely ignore the online signal from
control. This was to work around the times when the control plane
determines that a node is not online, while the node is still able to
connect to its peers. However, this breaks high-availability failover,
since nodes that are actually offline are never treated that way.

This PR makes nodeBackend.PeerIsReachable actually test reachability
by disco-pinging the peer node.

In the future, the client itself should collect reachability data from
active Wireguard flows.

Updates #17366
Updates tailscale/corp#30379
Updates tailscale/corp#32686
Updates tailscale/corp#33033

Signed-off-by: Simon Law <sfllaw@tailscale.com>

bradfitz commented on ipn/ipnlocal/node_backend.go:
```diff
@@ -378,20 +383,54 @@ func (nb *nodeBackend) PeerIsReachable(ctx context.Context, p tailcfg.NodeView)
 }
 
-func (nb *nodeBackend) peerIsReachable(ctx context.Context, p tailcfg.NodeView) bool {
+func (nb *nodeBackend) peerIsReachable(ctx context.Context, p tailcfg.NodeView) (bool, error) {
```
document what (false, nil) means and how that differs from (false, context expired) ?

sfllaw committed:
ipn/ipnlocal: extract ping implementation into its own file

There’s enough code to pull this out into its own ping.go file, since
local.go contains a lot of stuff.

The next commit will add a Pinger interface to this file, so I am
breaking out this refactor to make identifying the changes and
reviewing the actual patch easier.

Signed-off-by: Simon Law <sfllaw@tailscale.com>

sfllaw committed:
ipn/ipnlocal: plumb context into suggestExitNode functions

PR #17367 deferred plumbing context.Context into the suggestExitNode
functions. This patch uses the LocalBackend’s context, which is valid
until LocalBackend.Shutdown is called.

Updates #17366

Signed-off-by: Simon Law <sfllaw@tailscale.com>

sfllaw commented on control/controlclient/auto.go:
```diff
@@ -582,7 +582,7 @@ func (c *Auto) SetTKAHead(headHash string) {
 
 // sendStatus can not be called with the c.mu held.
-func (c *Auto) sendStatus(who string, err error, url string, nm *netmap.NetworkMap) {
+func (c *Auto) sendStatus(ctx context.Context, who string, err error, url string, nm *netmap.NetworkMap) {
```
I started to pass this context down to SetControlClientStatus, since that calls SuggestExitNode, but I haven’t figured out whether it’s OK for that to run until completion or whether it should be cancellable. I think it might be better to queue up SuggestExitNode and its matching RefreshExitNode as an eventbus listener? cc: @creachadair and @cmol?

sfllaw commented:
Abandoned in favour of #19639 which is a later, more mature, attempt.
